### PR TITLE
[5.2] Big performance increase for has() queries

### PR DESCRIPTION
**Pull request by acasar**
As you know, has() queries in Eloquent are really slow on big data sets. This has been discussed at length in the past: https://github.com/laravel/framework/issues/8720 https://github.com/laravel/framework/issues/3543 https://github.com/laravel/framework/issues/5328

We even had a PR, but it was quite "hacky": https://github.com/laravel/framework/pull/8748

But in fact there is a simple solution. Instead of:

``` sql
where (select count(*) from table) >= 1
where (select count(*) from table) < 1
```

We can simply run:

``` sql
where exists(select * from table)
where not exists(select * from table)
```

Of course, if we are doing anything else than `>=1` or `<1` the behaviour remains unchanged.

I had to add a new parameter (`$columns`) to `getRelationCountQuery` on all relation classes  to make it work, but it's actually a pretty simple change. In my tests i managed to get between 5X and 60X faster performance. If you have lots of data in the related table, the performance increase will be much more noticable.

Oh, and happy new year everyone :) :fireworks:

### CODE REVIEW

**Comment by taylorotwell** [comment]
I guess the database engines don't **actually** select the whole table in this case and just quit after the first row is found?

**Comment by acasar** [comment]
Exactly. After the first match it doesn't select any more rows.

**Comment by acasar** [comment]
And database engines also ignore the select part of the query. So ``select 1`` is basically the same as ``select *``.

**Comment by pulkitjalan** [comment]
could this be added to `5.1` also?

**Comment by GrahamCampbell** [comment]
> I can submit a pr?

No, sorry. 5.1 is only getting bug-fixes now.

**Comment by JosephSilber** [comment]
Why not use `select 1` explicitly?

**Comment by acasar** [comment]
It really doesn't make any difference in performance since ``select`` is not even evaluated, but if you prefer to be explicit, I can easily add it.

**Comment by JosephSilber** [comment]
Oh I see it's using the existing query. Yeah. No need to bother with that.

**Comment by CurosMJ** [comment]
Why even have a 'exists' clause in the SQL, what if we just parsed the result of 'select' query only?

**Comment by acasar** [comment]
@CurosMJ I'm not sure what you mean. This is for has() queries like: ``User::has('posts')->get()``.

**Comment by CurosMJ** [comment]
Instead of running 
```sql 
where exists(select 1 from table)
```

Why not just run 
```sql
select 1 from table
```
and see if that returned a row or not.

**Comment by acasar** [comment]
The full query is:
```sql
select * from users where exists (select * from posts where posts.user_id = post.id)
```
How would you rewrite that?

**Comment by CurosMJ** [comment]
sorry I misinterpreted the context. :stuck_out_tongue:

**Comment by thomthom** [comment]
This would improve the performance of `whereHas` as well as `has`?

**Comment by acasar** [comment]
Yeah it applies to all variants oh has().

**Comment by SpaceK33z** [comment]
I think this is a very good improvement. A while ago we noticed our API was very slow, and after debugging for a while it appeared to be the `has()` query. Rewriting it to a whereExists query had a great impact on the performance.

File: src/Illuminate/Database/Eloquent/Builder.php [hunk 1]
```diff
@@ -815,8 +854,6 @@ protected function mergeModelDefinedRelationWheresToHasQuery(Builder $hasQuery,
         $hasQuery->mergeWheres(
             $relationQuery->wheres, $relationQuery->getBindings()
         );
-
-        $this->query->addBinding($hasQuery->getBindings(), 'where');
```
Comment (taylorotwell) [inline src/Illuminate/Database/Eloquent/Builder.php]: Why was this removed?

Comment (acasar) [inline src/Illuminate/Database/Eloquent/Builder.php]: Just moved to another function.
